File: JavaScript/ExampleTests/common/StatusString.js
import * as tokens from "./tokens.js";

export class StatusString {
  static DNC_STRING = "?DNC?";

  constructor(code = "") {
    this.code = code;
  }

  static fromList(values) {
    const v = Array.from(values);
    return new StatusString(
      v[0] ?? ""
    );
  }

  /** Builds from the text form, e.g. Money as "25 USD". */
  static fromText(text) {
    const parts = tokens.require_(text, 1, "Status");
    return new StatusString(
      parts[0]
    );
  }

  toString() {
    return tokens.token(this.code);
  }

  equals(other) {
    if (!(other instanceof StatusString)) return false;
    return (this.code === StatusString.DNC_STRING || other.code === StatusString.DNC_STRING || this.code === other.code);
  }

  static equalLists(a, b) {
    if (a.length !== b.length) return false;
    return a.every((row, i) => row.equals(b[i]));
  }
}

File: Go/ExampleTests/common/status_string.go
package common

type StatusString struct {
	Code string
}

func NewStatusStringFromSlice(v []string) StatusString {
	s := StatusString{}
	if len(v) > 0 { s.Code = v[0] }
	return s
}

// NewStatusStringFromText builds from the text form, e.g. Money as "25 USD".
func NewStatusStringFromText(text string) StatusString {
	parts := RequireTokens(text, 1, "Status")
	return StatusString{
		Code: parts[0],
	}
}

func (s StatusString) String() string {
	return Token(s.Code)
}

func (s StatusString) Equals(o StatusString) bool {
	return DNCEqual(s.Code, o.Code)
}

func EqualStatusStringSlices(a, b []StatusString) bool {
	if len(a) != len(b) { return false }
	for i := range a {
		if !a[i].Equals(b[i]) { return false }
	}
	return true
}
